File: backend/product-service/src/main/java/com/optica/product/service/ProductService.java
package com.optica.product.service;

import com.optica.product.dto.ProductRequest;
import com.optica.product.dto.ProductResponse;
import com.optica.product.exception.ApiException;
import com.optica.product.model.Category;
import com.optica.product.model.Product;
import com.optica.product.repository.CategoryRepository;
import com.optica.product.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;
    private final CategoryRepository categoryRepository;

    public List<ProductResponse> list(Long categoryId) {
        var entities = (categoryId == null)
                ? productRepository.findAllByOrderByIdDesc()
                : productRepository.findByCategoryIdOrderByIdDesc(categoryId);
        return entities.stream().map(ProductResponse::from).toList();
    }

    public ProductResponse getById(Long id) {
        Product p = productRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Producto no encontrado"));
        return ProductResponse.from(p);
    }

    public ProductResponse create(ProductRequest req) {
        Category category = categoryRepository.findById(req.getCategoryId())
                .orElseThrow(() -> new ApiException(HttpStatus.BAD_REQUEST, "La categoria no existe"));

        Product saved = productRepository.save(Product.builder()
                .name(req.getName())
                .description(req.getDescription())
                .price(req.getPrice())
                .stock(req.getStock())
                .category(category)
                .build());
        return ProductResponse.from(saved);
    }

    public ProductResponse update(Long id, ProductRequest req) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Producto no encontrado"));

        Category category = categoryRepository.findById(req.getCategoryId())
                .orElseThrow(() -> new ApiException(HttpStatus.BAD_REQUEST, "La categoria no existe"));

        product.setName(req.getName());
        product.setDescription(req.getDescription());
        product.setPrice(req.getPrice());
        product.setStock(req.getStock());
        product.setCategory(category);

        return ProductResponse.from(productRepository.save(product));
    }

    public void delete(Long id) {
        if (!productRepository.existsById(id)) {
            throw new ApiException(HttpStatus.NOT_FOUND, "Producto no encontrado");
        }
        productRepository.deleteById(id);
    }
}

File: backend/product-service/src/services/productService.js
const productRepository = require('../repositories/productRepository');
const categoryRepository = require('../repositories/categoryRepository');
const { toProductResponseDTO } = require('../dtos/productDtos');

class ProductService {
  async list() {
    const products = await productRepository.findAll();
    return products.map(toProductResponseDTO);
  }

  async listByCategory(categoryId) {
    const category = await categoryRepository.findById(categoryId);
    if (!category) {
      const err = new Error('Categoria no existe');
      err.status = 404;
      throw err;
    }
    const products = await productRepository.findByCategory(categoryId);
    return products.map(toProductResponseDTO);
  }

  async getById(id) {
    const product = await productRepository.findById(id);
    if (!product) {
      const err = new Error('Producto no encontrado');
      err.status = 404;
      throw err;
    }
    return toProductResponseDTO(product);
  }

  async create(data) {
    const category = await categoryRepository.findById(data.category_id);
    if (!category) {
      const err = new Error('La categoria no existe');
      err.status = 400;
      throw err;
    }
    const created = await productRepository.create(data);
    const full = await productRepository.findById(created.id);
    return toProductResponseDTO(full);
  }

  async update(id, data) {
    if (data.category_id) {
      const category = await categoryRepository.findById(data.category_id);
      if (!category) {
        const err = new Error('La categoria no existe');
        err.status = 400;
        throw err;
      }
    }
    const updated = await productRepository.update(id, data);
    if (!updated) {
      const err = new Error('Producto no encontrado');
      err.status = 404;
      throw err;
    }
    return toProductResponseDTO(updated);
  }

  async remove(id) {
    const ok = await productRepository.delete(id);
    if (!ok) {
      const err = new Error('Producto no encontrado');
      err.status = 404;
      throw err;
    }
    return { deleted: true };
  }
}

module.exports = new ProductService();
